README.md
# Telegram Decoder 

![](https://img.shields.io/endpoint?url=https%3A%2F%2Fhydra.hq.c3d2.de%2Fjob%2Ftlm-solutions%2Ftelegram-decoder%2Ftelegram-decoder.x86_64-linux%2Fshield)
![built with nix](https://builtwithnix.org/badge.svg)


**Contact:** <hello@tlm.solutions>

This service takes the raw bit stream from the gnuradio pipeline and tries to decode valid telegrams
from it. This service implement the full CRC and decodation of **R09.16** all other telegram variants
are sent as raw telegrams. All decoded telegram are sent to [data-accumulator](https://github.com/tlm-solutions/data-accumulator).


## Service Configuration 


### Environment Variables

- **AUTHENTICATION_TOKEN_PATH** per default set to `/etc/telegram-decoder/token`

### Config File

the config flag in the command line inputs points to a file following the pattern described below.

```
{
    "name": "Station Name",
    "lat": -1.0,
    "lon": -1.0,
    "id": "<YOUR STATION ID>",
    "region": -1
}
```



CHANGELOG.md
# Changelog

## Unreleased

### Breaking

### Added

### Fixed

### Misc

## 0.3.0
This release is a start of us properly releasing. Anything before this lost to
history.


src/decoder/test.rs
use crate::decoder::Decoder;
use tlms::{locations::R09Types, telegrams::r09::R09Telegram};
use serde::Deserialize;
use std::collections::hash_map::DefaultHasher;
use std::hash::Hash;
use std::hash::Hasher;
use std::sync::mpsc;
use std::sync::mpsc::SyncSender;

extern crate derive_builder;

#[derive(Deserialize)]
struct TelegramFrame {
    zv: u32,
    zw: u32,
    mp: u32,
    pr: u32,
    ha: u32,
    ln: u32,
    kn: u32,
    zn: u32,
    zl: u32,
    junction: u32,
    junction_number: u32,
    request_status: u32,
}

#[derive(Deserialize)]
struct ValidR09_16Telegram {
    input: Vec<u8>,
    output: TelegramFrame,
}

pub fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}

macro_rules! decode_telegrams_from_file {
    ($file: expr, $disable_error_correction: expr) => {{
        let (sender_r09, receiver_r09) = mpsc::sync_channel(1);
        let mut senders_r09: Vec<SyncSender<R09Telegram>> = Vec::new();
        senders_r09.push(sender_r09);

        let mut decoder = Decoder::new(&senders_r09, &Vec::new(), $disable_error_correction);

        const FILE_STR: &'static str = include_str!($file);

        let parsed: Vec<ValidR09_16Telegram> =
            serde_json::from_str(&FILE_STR).expect("JSON was not well-formatted");

        for (i, item) in parsed.iter().enumerate() {
            let telegram = &item.output;

            let expected_telegram = R09Telegram {
                telegram_type: R09Types::R16,
                delay: Some((telegram.zv as i32 * -2i32 + 1i32) * (telegram.zw as i32)),
                reporting_point: telegram.mp,
                junction: telegram.junction,
                direction: telegram.junction_number as u8,
                request_status: telegram.request_status as u8,
                priority: Some(telegram.pr as u8),
                direction_request: Some(telegram.ha as u8),
                line: Some(telegram.ln),
                run_number: Some(telegram.kn),
                destination_number: Some(telegram.zn),
                train_length: Some(telegram.zl as u8),
                vehicle_number: None,
                operator: None,
            };

            decoder.process(&item.input.as_ref());
            let received_telegram = receiver_r09.recv().unwrap();

            assert_eq!(
                calculate_hash(&received_telegram),
                calculate_hash(&expected_telegram)
            );

            println!("{}/{} OK", i + 1, parsed.len());
        }
    }};
}

#[test]
fn test_decode_valid_r09_16_telegrams() {
    decode_telegrams_from_file!("../../data/valid_r09_16_telegrams.json", true);
}

#[test]
fn test_decode_1bit_error_r09_16_telegrams() {
    decode_telegrams_from_file!("../../data/1bit_error_r09_16_telegrams.json", false);
}


src/main.rs
extern crate serde_derive;

mod decoder;
mod structs;
mod sink;

// modules
use decoder::Decoder;
use structs::Args;
use sink::{DataSinkConfig, send_r09, send_raw};

use tlms::{
    receivers::RadioReceiver,
    telegrams::{
        r09::{
            R09Telegram,
        },
        raw::{
            RawTelegram,
        }
    },
};

// extern creates
use clap::Parser;
use env_logger;
use log::info;

// standard lib
use std::thread;
use std::fs::read_to_string;
use std::net::UdpSocket;
use std::sync::mpsc;
use std::sync::mpsc::{Receiver, SyncSender};

fn main() {
    env_logger::init();

    const BUFFER_SIZE: usize = 2048;
    let args = Args::parse();

    let stop_mapping = String::from(&args.config);
    let contents = read_to_string(stop_mapping).expect("Something went wrong reading the file");

    let station_config: RadioReceiver =
        serde_json::from_str(&contents).expect("JSON was not well-formatted");
    
    let mut senders_r09: Vec<SyncSender<R09Telegram>> = Vec::new(); 

    let mut senders_raw: Vec<SyncSender<RawTelegram>> = Vec::new(); 

    for server in &args.server {
        let r09_sink_config = DataSinkConfig::new(args.offline, &server, station_config.clone());
        let raw_sink_config = r09_sink_config.clone();

        let (sender_r09, mut receiver_r09) = mpsc::sync_channel(50);
        senders_r09.push(sender_r09);

        let _thread_send_r09 = thread::spawn(move || {
            send_r09(&mut receiver_r09, &r09_sink_config.clone());
        });

        let (sender_raw, mut receiver_raw) = mpsc::sync_channel(50);
        senders_raw.push(sender_raw);

        let _thread_send_raw = thread::spawn(move || {
            send_raw(&mut receiver_raw, &raw_sink_config.clone());
        });
    }

    let mut decoder = Decoder::new(&senders_r09, &senders_raw, args.disable_error_correction);

    info!("Starting DVB Dump Telegram Decoder ... ");
    if args.disable_error_correction {
        info!("Error correction is disabled ...");
    }

    let addr = format!("{}:{}", &args.host, &args.port);
    let socket = UdpSocket::bind(addr).unwrap();
    let (tx, rx): (SyncSender<[u8; BUFFER_SIZE]>, Receiver<[u8; BUFFER_SIZE]>) =
        mpsc::sync_channel(400);

    let _thread_decode = thread::spawn(move || {
        loop {
            let data = rx.recv().unwrap();
            decoder.process(&data);
        }
    });

    loop {
        let mut buffer = [0; BUFFER_SIZE];
        let (_amt, _src) = socket.recv_from(&mut buffer).unwrap();
        tx.send(buffer.clone()).unwrap();
    }
}


src/sink.rs
use tlms::{
    telegrams::r09::{R09Telegram, R09ReceiveTelegram}, 
    telegrams::raw::{RawTelegram, RawReceiveTelegram},
    receivers::RadioReceiver, 
    telegrams::AuthenticationMeta
};
use std::time::Duration;
use std::sync::mpsc::Receiver;
use log::{warn, error, debug};
use std::env;

use chrono::Utc;
use reqwest::blocking::Client;

#[derive(Clone, Debug)]
pub struct DataSinkConfig {
    pub token: String,
    pub host: String,
    pub station: RadioReceiver
}

impl DataSinkConfig {
    pub fn new(offline: bool, host: &String, station: RadioReceiver) -> DataSinkConfig {
        let token: String;
        if offline {
            debug!("offline mode no token required");
            token = String::from("");
        } else {
            debug!("reading token from {:?}", env::var("AUTHENTICATION_TOKEN_PATH"));

            token = env::var("AUTHENTICATION_TOKEN_PATH")
                .map(|token_path| {
                    String::from_utf8_lossy(&std::fs::read(token_path).unwrap())
                        .parse::<String>()
                        .unwrap()
                })
                .unwrap()
                .lines()
                .next()
                .unwrap()
                .to_string();

        }
        DataSinkConfig {
            token: token,
            host: host.clone(),
            station: station
        }
    }
}

pub fn send_r09(r09_receiver: &mut Receiver<R09Telegram>, sink: &DataSinkConfig) {
    let client = Client::new();

    loop {
        let wrapped_telegram = r09_receiver.recv();
        match  wrapped_telegram {
            Ok(telegram) => {
                let auth = AuthenticationMeta {
                    station: sink.station.id.clone(),
                    token: sink.token.clone(),
                    time: Utc::now().naive_utc()
                };

                let url = format!("{}/telegram/r09", &sink.host);
                match client
                    .post(&url)
                    .timeout(Duration::from_millis(750))
                    .json(&R09ReceiveTelegram {
                        auth: auth.clone(),
                        data: telegram.clone(),
                    })
                    .send()
                {
                    Err(_) => {
                        warn!("Connection Timeout! {}", &sink.host);
                    }
                    _ => {}
                }
            }
            Err(e) => {
                error!("received following error from r09 pipeline {:?}", e);
            }
        }
    }
}

pub fn send_raw(raw_receiver: &mut Receiver<RawTelegram>, sink: &DataSinkConfig) {
    let client = Client::new();

    loop {
        let wrapped_telegram = raw_receiver.recv();
        match wrapped_telegram {
            Ok(telegram) => {
                let auth = AuthenticationMeta {
                    station: sink.station.id.clone(),
                    token: sink.token.clone(),
                    time: Utc::now().naive_utc()
                };

                let url = format!("{}/telegram/raw", &sink.host);
                match client
                    .post(&url)
                    .timeout(Duration::from_millis(750))
                    .json(&RawReceiveTelegram {
                        auth: auth.clone(),
                        data: telegram.clone(),
                    })
                    .send()
                {
                    Err(_) => {
                        warn!("Connection Timeout! {}", &sink.host);
                    }
                    _ => {}
                }
            }
            Err(e) => {
                error!("received following error from raw pipeline {:?}", e);
            }
        }
    }
}


src/structs.rs
extern crate derive_builder;

use clap::Parser;

#[derive(Parser, Debug)]
#[clap(name = "telegram-decoder")]
#[clap(author = "hello@tlm.solutions")]
#[clap(version = env!("CARGO_PKG_VERSION"))]
#[clap(about = "Extracts telegram from raw bit streams and sends them.", long_about = None)]
pub struct Args {
    #[clap(short, long, default_value_t = String::from("127.0.0.1"))]
    pub host: String,

    #[clap(short, long, default_value_t = 40000)]
    pub port: u32,

    #[clap(short, long)]
    pub server: Vec<String>,

    #[clap(short, long, default_value_t = String::from("config.json"))]
    pub config: String,

    #[clap(short, long, action)]
    pub offline: bool,

    #[clap(short, long, action)]
    pub disable_error_correction: bool,

    #[clap(short, long, action)]
    pub verbose: bool,
}


src/telegram.rs
use std::convert::TryFrom;

enum PackageType {
    StandardResponse = 0,
    AbsoluteLocation = 1,
    MessageAcknowledgements = 2,
    Load = 3,
    RouteRun = 4,
    DriversNo = 5,
    NumberOfTrailer = 6,
    DestinationNumber = 7,
    UnitSerialNumber = 8,
    ApplicationSpecific = 9,
    Spare = 10,
    ReducedMessage = 10,
    StandardMessage = 11,
    StandardMessagePriority = 12,
    StandardMessageLineNumber = 13,
    StandardMessageLineNumberRunNumber = 14,
    NotUsed = 15,
    MaximumMessage = 16
}


impl TryFrom<u8> for PackageType {
    type Error = ();

    fn try_from(v: i32) -> Result<Self, Self::Error> {
        match v {
            x if x == PackageType::StandardMessage as u8 => Ok(PackageType::StandardMessage),
            x if x == PackageType::AbsoluteLocation as u8 => Ok(PackageType::AbsoluteLocation),
            x if x == PackageType::MessageAcknowledgements as u8 => Ok(PackageType::MessageAcknowledgements),

            x if x == PackageType::Load as u8 => Ok(PackageType::Load),
            x if x == PackageType::RouteRun as u8 => Ok(PackageType::RouteRun),
            x if x == PackageType::DriversNo as u8 => Ok(PackageType::DriversNo),
            x if x == PackageType::NumberOfTrailer as u8 => Ok(PackageType::NumberOfTrailer),
            x if x == PackageType::DestinationNumber as u8 => Ok(PackageType::DestinationNumber),
            x if x == PackageType::UnitSerialNumber as u8 => Ok(PackageType::UnitSerialNumber),
            x if x == PackageType::ApplicationSpecific as u8 => Ok(PackageType::ApplicationSpecific),
            x if x == PackageType::Spare as u8 => Ok(PackageType::Spare),
            x if x == PackageType::ReducedMessage as u8 => Ok(PackageType::ReducedMessage),
            x if x == PackageType::StandardMessage as u8 => Ok(PackageType::StandardMessage),
            x if x == PackageType::StandardMessagePriority as u8 => Ok(PackageType::StandardMessagePriority),
            x if x == PackageType::StandardMessageLineNumber as u8 => Ok(PackageType::StandardMessageLineNumber),
            x if x == PackageType::StandardMessageLineNumberRunNumber as u8 => Ok(PackageType::StandardMessageLineNumberRunNumber),
            x if x == PackageType::NotUsed as u8 => Ok(PackageType::NotUsed),
            x if x == PackageType::MaximumMessage as u8 => Ok(PackageType::MaximumMessage),
            x if x == MyEnum::C as i32 => Ok(MyEnum::C),
            _ => Err(()),
        }
    }
}



src/decoder/mod.rs
pub mod structs;
#[cfg(test)]
mod test;

pub use structs::parse_r09_telegram;

use tlms::telegrams::{r09::R09Telegram, raw::RawTelegram, TelegramType};

use g2poly::G2Poly;
use log::info;
use num_traits::cast::FromPrimitive;

use std::collections::HashMap;
use std::sync::mpsc::SyncSender;

struct RepairedTelegram {
    data: Vec<u8>,
    number_of_bits_repaired: u32,
}

pub struct Decoder {
    //server: Vec<String>,
    //station_config: RadioReceiver,
    maps: Vec<HashMap<u64, Vec<u8>>>,
    //token: String,
    r09_senders: Vec<SyncSender<R09Telegram>>,
    raw_senders: Vec<SyncSender<RawTelegram>>,
}

impl Decoder {
    pub fn new(
        r09_senders: &Vec<SyncSender<R09Telegram>>,
        raw_senders: &Vec<SyncSender<RawTelegram>>,
        disable_error_correction: bool,
    ) -> Decoder {
        let mut maps: Vec<HashMap<u64, Vec<u8>>> = Vec::new();

        for len in 5..22 {
            let mut map: HashMap<u64, Vec<u8>> = HashMap::new();

            if !disable_error_correction {
                // 1 bit errors
                for i in 0..(len * 8) {
                    let mut data = vec![0u8; len];
                    let idx = (i / 8) as usize;
                    let pos = i % 8;
                    data[idx] ^= 1 << pos;

                    let value: u64 = Decoder::crc16_remainder(&data).0;

                    if let Some(_) = map.get(&value) {
                        assert!(false);
                    } else {
                        map.insert(value, data);
                    }
                }

                // 2 bit errors
                for i in 0..(len * 8) {
                    for j in (i + 1)..(len * 8) {
                        let mut data = vec![0u8; len];
                        let idx = (i / 8) as usize;
                        let pos = i % 8;
                        data[idx] ^= 1 << pos;
                        let idx = (j / 8) as usize;
                        let pos = j % 8;
                        data[idx] ^= 1 << pos;

                        let value: u64 = Decoder::crc16_remainder(&data).0;

                        if let Some(_) = map.get(&value) {
                            assert!(false);
                        } else {
                            map.insert(value, data);
                        }
                    }
                }

                // 3 bit errors
                // this might work sometimes...
                // so the algorithm for creating the map is a bit more complicated
                if len < 14 {
                    let mut blacklist: Vec<u64> = Vec::new();
                    let mut map_3bit: HashMap<u64, Vec<u8>> = HashMap::new();

                    for i in 0..(len * 8) {
                        for j in (i + 1)..(len * 8) {
                            for k in (j + 1)..(len * 8) {
                                let mut data = vec![0u8; len];
                                let idx = (i / 8) as usize;
                                let pos = i % 8;
                                data[idx] ^= 1 << pos;
                                let idx = (j / 8) as usize;
                                let pos = j % 8;
                                data[idx] ^= 1 << pos;
                                let idx = (k / 8) as usize;
                                let pos = k % 8;
                                data[idx] ^= 1 << pos;

                                let value: u64 = Decoder::crc16_remainder(&data).0;

                                // only try to add it, if the value is not allready corrected by 1 or 2 bit
                                // error correction
                                if None == map.get(&value) {
                                    if let Some(_) = map_3bit.get(&value) {
                                        // throw out the value if it occurs multiple times (hamming
                                        // distance too low)
                                        map_3bit.remove(&value);
                                        blacklist.push(value);
                                    } else {
                                        // add the value if it is not in the blacklist (hamming distance
                                        // too low)
                                        if blacklist.iter().all(|&v| v != value) {
                                            map_3bit.insert(value, data);
                                        }
                                    }
                                }
                            }
                        }
                    }

                    // extend the map with 3 bit error correction values
                    map.extend(map_3bit);
                }
            }

            maps.push(map)
        }

        Decoder {
            maps: maps,
            r09_senders: r09_senders.clone(),
            raw_senders: raw_senders.clone(),
        }
    }

    pub fn process(&mut self, data: &[u8]) {
        let data_copy = data.clone();

        self.decode(data_copy);
    }

    pub fn decode(&mut self, data: &[u8]) {
        // minimum message size is 3 bytes + 2 bytes crc
        const MINIMUM_SIZE: usize = 5;
        // C09 fixed size of 4 bytes + variable length 4 bits (15 bytes) + 2 bytes CRC
        // max length 4 + 15 + 2 = 21
        const MAXIMUM_SIZE: usize = 21;

        let mut byte_array: Vec<u8> = Vec::new();

        for i in 0..MAXIMUM_SIZE {
            if (i + 1) * 9 - 1 >= data.len() {
                break;
            }
            byte_array.push(Decoder::bit_to_bytes(&data[i * 9..(i + 1) * 9 - 1]));
        }

        //let mut collection: Vec<Box<dyn AbstractTelegram>> = Vec::new();

        // Abort if we don't have enough data for a packet
        if byte_array.len() < MINIMUM_SIZE {
            return;
        }

        // try decoding for every possible length
        for telegram_length in MINIMUM_SIZE..(byte_array.len() - 1) {
            let mut telegram_array = Vec::new();
            telegram_array.extend_from_slice(&byte_array[0..telegram_length]);

            // invert crc
            telegram_array[telegram_length - 2] ^= 0xff;
            telegram_array[telegram_length - 1] ^= 0xff;

            let rem = Decoder::crc16_remainder(&telegram_array);

            let mut telegrams: Vec<RepairedTelegram> = Vec::new();

            if rem == G2Poly(0) {
                // no errors, decode
                telegrams.push(RepairedTelegram {
                    data: (&telegram_array[0..(telegram_length - 2)]).to_vec(),
                    number_of_bits_repaired: 0u32,
                });
            } else {
                // errors. try to fix them
                if let Some(error) = self.maps[telegram_length - MINIMUM_SIZE].get(&rem.0) {
                    assert_eq!(error.len(), telegram_length);

                    let mut repaired_telegram = telegram_array.clone();
                    for i in 0..error.len() {
                        repaired_telegram[i] ^= error[i];
                    }

                    assert_eq!(
                        Decoder::crc16_remainder(&repaired_telegram),
                        G2Poly(0)
                    );

                    telegrams.push(RepairedTelegram {
                        data: (&repaired_telegram[0..(telegram_length - 2)]).to_vec(),
                        number_of_bits_repaired: error.iter().map(|x| x.count_ones()).sum(),
                    });
                }
            }

            for telegram in telegrams {
                self.parse_telegram(&telegram)
            }
        }
    }

    // data is a vector of data without crc
    // TODO: change this into a vector. There is the possibilty for a valid R09 telegram being a
    // valid C09 telegram and vice versa.
    fn parse_telegram(&mut self, repair_telegram: &RepairedTelegram) {
        let mode: u8 = repair_telegram.data[0] >> 4;
        let length: usize = repair_telegram.data.len();

        // length has to be at least 3 bytes
        if length < 3 {
            return;
        }

        // these modes may only have length 3 (R telegrams) or 4 (C telegrams)
        // lower bound is already checked above
        if mode <= 4 || mode >= 10 {
            if length > 4 {
                return;
            }
        }

        if mode == 9 {
            // parse R09.x
            if 3 + (repair_telegram.data[1] & 0xf) as usize == length {
                return self.parse_r09(&repair_telegram.data);
            }
            // parse C09.x
            if 4 + (repair_telegram.data[2] & 0xf) as usize == length {
                let c09_type = repair_telegram.data[2] >> 4;
                let c09_length = repair_telegram.data[2] & 0xf;

                // TODO
                info!("[!] Recevied C09.{}.{}", c09_type, c09_length);

                return;
            }

            return;
        } else {
            if repair_telegram.number_of_bits_repaired > 0 {
                return;
            }
            let telegram_type = TelegramType::from_u8(mode).unwrap();

            let raw_telegram = RawTelegram {
                telegram_type: telegram_type,
                data: repair_telegram.data.clone(),
            };

            info!("Detected RawTelegram: {:?}", raw_telegram);
            for sender in &self.raw_senders {
                let _ = sender.clone().try_send(raw_telegram.clone());
            }
        }

        // We removed the one variable length telegrams of the R-series R09, others are 3 bytes
        // long.
        // The C-series has C09, which is variable length, but we don't know anything about C05-C08.
        // They are probably only 4 bytes long, like other ones from the C-series, but we don't
        // know.
        match length {
            3 => info!("[!] Received R {}", mode),
            _ => info!("[!] Received C {}", mode),
        };
    }

    fn parse_r09(&mut self, data: &[u8]) {
        // lower nibble of the mode
        let r09_type = data[0] & 0xf;
        let r09_length = data[1] & 0xf;

        assert_eq!(3 + r09_length as usize, data.len());

        // decode R09.1x
        if r09_type == 1 {
            // TODO: if BCD is not BCD, throw it out
            match parse_r09_telegram(data) {
                Some(data) => {
                    info!("[!] Received R09.{}.{}: {:?}", r09_type, r09_length, data);
                    for sender in &self.r09_senders {
                        let _ = sender.clone().try_send(data.clone());
                    }
                }
                None => {
                    if !(r09_length == 4 || r09_length == 6) {
                        info!("[!] Received R09.{}.{}: {:?}", r09_type, r09_length, data);
                    }
                }
            }
        } else {
            info!("[!] Received R09.{}.{}: {:?}", r09_type, r09_length, data);
        }
    }

    // converts a list of bits into a single byte
    fn bit_to_bytes(data: &[u8]) -> u8 {
        if data.len() != 8 {
            return 0;
        } else {
            let mut byte_value: u8 = 0;

            for bit in 0..8u8 {
                if data[bit as usize] > 0 {
                    byte_value += 1 << bit;
                }
            }

            return byte_value;
        }
    }

    fn crc16_remainder(data: &Vec<u8>) -> G2Poly {
        const POLY: G2Poly = G2Poly(0x16f63);
        const ALPHA: G2Poly = G2Poly(0b10);

        let mut rem = G2Poly(0);

        for i in 0..data.len() {
            for j in 0..8 {
                let offset: u64 = (i * 8).try_into().unwrap();
                if 1 == (data[data.len() - i - 1] >> (7 - j)) & 0x1 {
                    rem = (rem + ALPHA.pow_mod(j + offset, POLY)) % POLY;
                }
            }
        }

        rem
    }
}


decode_telegrams.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import requests
import socket
import crcmod
import sys
import time

host = "academicstrokes.com"

anmeldung_typ = {
        0: "Vor",
        1: "An",
        2: "Ab",
        3: "Tür zu"
        }

def get_knotenpunkt(nummer):
    kpdict = {
            # AP 3.1.1(.2) DDL003
            102: "LSA Anton-/Leipziger Straße",
            # AP 3.2.1 DDL072
            103: "Anton-/Zur Eisenbahnstraße",
            # Bf Neustadt
            # AP 3.3.1(.2) DDL071
            104: "Schlesischer Platz",
            # AP 3.4.1.1
            107: "Albertplatz",
            108: "Albertplatz West",
            109: "Bautzner/Rothenburger",
            # AP 3.10.1(.2) DDL036
            119: "Straßburger Platz",
            # AP 3.19.1 DDL310
            120: "Lennéstraße/Hauptallee",
            121: "Lennéplatz",
            136: "Könneritzstraße/Jahnstraße",

            200: "Synagoge",
            # AP 3.3.1.1
            # ???
            201: "St. Petersburger Straße/Pulmann-Newa",
            203: "Pirnaischer Platz",
            # AP 3.6.1.1
            204: "Rathenauplatz",
            # AP 3.5.1.1
            205: "Carolaplatz",
            208: "Bodenbacher / Zwinglistraße",
            268: "Albertplatz Süd",

            324: "Fetscherplatz",
            346: "Schillerplatz",
            363: "Blasewitzer /Fetscherstraße",

            421: "Wasaplatz",
            429: "Moränenende/Breitscheidstraße",
            435: "Moränenende/Wilhelm-Liebknecht-Straße",
            459: "Mügelner Straße/Moränenende",

            504: "Reichenbachstraße",
            506: "Fritz-Foerster-Platz",

            604: "Nürnberger /Budapester Str.",
            606: "Chemnitzer /Nöthritzer Str.",
            651: "Nürnberger Platz",

            804: "Tharandter /Kesselsdorfer Straße",
            # AP 3.35.1.2 DDL069
            851: "Löbtauer Str. / Fröbelstraße",

            1101: "Louisenstraße",
            1102: "Bischofsweg",
            1103: "Tannenstraße",
            1104: "Staufenbergallee",
            1151: "S-Bahnhof Bischofsplatz",

            1202: "LSA BTZ <-> Pulsnitzer",
            1203: "Pulsnitzer Straße",
            1205: "Diakonissenkrankenhaus",
            1220: "Nordstraße",

            1301: "Körnerplatz/Grundstr.",
            # AP 3.4.1(.2) DDL184

            1404: "Liststraße",

            1501: "Hansastraße/Eisenbahnstraße",
            1504: "Friedensstraße"
            }

    try:
        return kpdict[nummer]
    except:
        return str(nummer)

def bitvec_to_bytes(vec):
    if len(vec) != 8:
        return 0

    a = 0
    for i in range(8):
        if vec[i] > 0:
            a += 1 << i

    return a

def decode(payload):
    try:
        # check minimum length 3 data bytes + 2 crc bytes
        _ = payload[3 + 2 - 1]

        crc_at = lambda at: (payload[at+1] << 8 | payload[at]) ^ 0xffff
        crc16_func = crcmod.mkCrcFun(0x16f63, rev=True, initCrc=0x0000, xorOut=0x0000)

        # find packets with unknown length by checking the crc over multiple sizes
        length = 0

        for at in range(4, len(payload) - 1):
            if crc_at(at) == crc16_func(bytearray(payload[0:at])):
                length = at
        # could not find proper length, discard
        if length == 0:
            return False
        mode = payload[0] >> 4
        r09_type = payload[0] & 0xf

        print("whuuu", payload, mode, r09_type)
        if mode == 9 and r09_type == 1:
            length = payload[1] & 0xf
            print(length, len(payload))
            # check length
            _ = payload[3 + length + 2 - 1]

            # check crc
            if crc_at(9) != crc16_func(bytearray(payload[0:9])):
                print("CRC Dont match")
                return False

            if length == 6:
                ZV = payload[1] >> 7
                ZW = (payload[1] >> 4) & 0x7
                MP = ((payload[2] >> 4) << 12) | ((payload[2] & 0x0f) << 8) | ((payload[3] >> 4) << 4) | ((payload[3] & 0x0f))
                PR = payload[4] >> 6
                HA = (payload[4] >> 4) & 0x3
                LN = 100 * (payload[4] & 0xf) + 10 * (payload[5] >> 4) +  (payload[5] & 0xf)
                KN = 10 * (payload[6] >> 4) + (payload[6] & 0xf)
                ZN = 100 * (payload[7] >> 4) + 10 * (payload[7] & 0xf) + (payload[8] >> 4)
                R = (payload[8] >> 3) & 0x1
                ZL = payload[8] & 0x7

                KnotenPunkt = (MP >> 2) // 10
                KnotenPunktNummer = (MP >> 2) - 10 * KnotenPunkt
                RequestStatus = MP & 0x3
                print(payload) 
                deviation = "{}{}:{}min".format(chr(0x2b+2*ZV), (ZW >> 1), (ZW & 1) * 30)
                # TODO, XXX this is wrong!
                print("Liniennummer: {}, Kursnummer: {}, Zielnummer: {}, t{}".format(LN, KN, ZN, deviation))
                print("ZV: {}, ZW: {}, MP: {} / {}, MP: {}, PR: {}, HA: {}, R: {}, ZL: {}".format(ZV, ZW, hex(MP >> 2), MP & 0x3, hex(MP), PR, HA, R, ZL))
                print("Knotenpunkt: {} / {} / {}".format(get_knotenpunkt(KnotenPunkt), KnotenPunktNummer, anmeldung_typ[RequestStatus]))

                req_payload = {
                    "time_stamp": int(time.time()),
                    "lat": 51.027107,
                    "lon": 13.723566,
                    "station_id": 100,
                    "line": LN,
                    "course_number": KN,
                    "destination_number": ZN,
                    "zv": ZV,
                    "zw": ZW,
                    "mp": MP,
                    "pr": PR,
                    "ha": HA,
                    "ln": LN,
                    "kn": KN,
                    "zn": ZN,
                    "r": R,
                    "zl": ZL,
                    "junction": KnotenPunkt,
                    "junction_number": KnotenPunktNummer,
                    "request_status": RequestStatus
                }

                r = requests.post("http://{}/formatted_telegram".format(host), json=req_payload)
                if r.status_code != 200:
                    print(f"Request yield {r.status_code}: ",req_payload)
            else:
                print("Mode: {}, Type: {}, Length: {}".format(mode, r09_type, length))
                print(payload)
                req_payload = {
                    "time_stamp": time.time(),
                    "lat": 51.027107,
                    "lon": 13.723566,
                    "station_id": 100,
                    "raw_data": payload
                }
                r = requests.post("http://{}/raw".format(host), json=req_payload)
                if r.status_code != 200:
                    print(f"Request yield {r.status_code}: ",req_payload)

        else:
            p = list(map(hex, payload))
            if mode == 9:
                print("Mode: {}, Type: {}, Length: {}, Payload: {}".format(mode, r09_type, length, p[:length+2]))
            else:
                if length > 4:
                    # discard
                    return False
                print("Mode: {}, Length: {}, Payload: {}".format(mode, length, p[:length+2]))

        return True

    except IndexError:
        pass

    return False

def one_bit_error_list(payload):
    corrected = []

    for i in range(len(payload)):
        for j in range(8):
            tmp = payload.copy()
            tmp[i] ^= (1 << j)
            corrected.append(tmp)

    return corrected

ip = "127.0.0.1"
port = 40000
sample_size = 100

# Create a UDP socket
s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
# Bind the socket to the port
server_address = (ip, port)
s.bind(server_address)

print("[*] Started Telegram Decoder")

while True:
    data, address = s.recvfrom(4096)

    i = 0
    bits = []
    while i < len(data):
        v = sum(data[i:i+2])
        if v != 1:
            bits.append(1)
        else:
            bits.append(0)
        i += 2

    # assume a max length of 20 bytes
    # 20 * 8 + 19 bits inbetween
    payload = []
    i = -1
    while i <= 180:
        payload.append(bitvec_to_bytes(data[i+1:i+9]))
        i += 9

    # print("======")
    # print(list(map(hex, payload)))

    return_code = decode(payload)
    # try to decode every 1 bit error
    one_bit_correted = False
    if return_code != True:
        for corrected in one_bit_error_list(payload):
            return_code = decode(corrected)
            if return_code == True:
                one_bit_correted = True
                break

    if one_bit_correted:
        print("one bit correction worked!!!")

    if return_code == True:
        print("======")
